File: Assets/Bundles/Client/Resources/TS/Controllers/Match/SpectateController.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Controller = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Controller
local ClientSignals = require("Client/TS/ClientSignals").ClientSignals
local OrbitCameraMode = require("Client/TS/Controllers/Global/Camera/DefaultCameraModes/OrbitCameraMode").OrbitCameraMode
local Entity = require("Shared/TS/Entity/Entity").Entity
local Game = require("Shared/TS/Game").Game
local Mouse = require("Shared/TS/UserInput/init").Mouse
local Bin = require("Shared/TS/Util/Bin").Bin
local SpectateController
do
	SpectateController = setmetatable({}, {
		__tostring = function()
			return "SpectateController"
		end,
	})
	SpectateController.__index = SpectateController
	function SpectateController.new(...)
		local self = setmetatable({}, SpectateController)
		return self:constructor(...) or self
	end
	function SpectateController:constructor(cameraController, entityController, bwController)
		self.cameraController = cameraController
		self.entityController = entityController
		self.bwController = bwController
		self.spectateCamDistance = 5
		self.spectateIndex = 0
		self.currentlySpectatingEntityId = 0
	end
	function SpectateController:OnStart()
		-- Start spectating once player is eliminated:
		ClientSignals.PlayerEliminated:Connect(function(event)
			if event.player ~= Game.LocalPlayer then
				return nil
			end
			self:StartSpectating()
		end)
	end
	function SpectateController:ObserveSpectatorTarget(callback)
		local cleanup
		local found = false
		local _value = self.currentlySpectatingEntityId
		if _value ~= 0 and (_value == _value and _value) then
			local entity = Entity:FindById(self.currentlySpectatingEntityId)
			if entity then
				found = true
				cleanup = callback(entity)
			end
		end
		if not found then
			cleanup = callback(nil)
		end
		local bin = Bin.new()
		bin:Add(ClientSignals.SpectatorTargetChanged:Connect(function(event)
			local _result = cleanup
			if _result ~= nil then
				_result()
			end
			cleanup = callback(event.entity)
		end))
		return function()
			bin:Clean()
		end
	end
	function SpectateController:StartSpectating()
		local entities = self:GetSortedEntities()
		if #entities == 0 then
			return nil
		end
		local bin = Bin.new()
		local orbitCamMode = OrbitCameraMode.new(entities[1].model.transform, self.spectateCamDistance)
		self.cameraController:SetMode(orbitCamMode)
		self:GoToIncrement(orbitCamMode, 0)
		-- Handle changing who is spectated:
		local mouse = bin:Add(Mouse.new())
		mouse.LeftDown:Connect(function()
			self:GoToIncrement(orbitCamMode, 1)
		end)
		mouse.RightDown:Connect(function()
			self:GoToIncrement(orbitCamMode, -1)
		end)
		bin:Connect(ClientSignals.EntitySpawn, function(event)
			self:FitIndexToId()
		end)
		bin:Connect(ClientSignals.EntityDespawn, function(entity)
			-- If currently-spectated entity goes away, switch to the next entity:
			if entity.id == self.currentlySpectatingEntityId then
				self:FitIndexToId()
				self:GoToIncrement(orbitCamMode, 1)
			else
				self:FitIndexToId()
			end
		end)
		-- Clean up after camera mode is changed:
		bin:Connect(self.cameraController.cameraSystem.ModeChangedBegin, function(newMode, oldMode)
			if oldMode == orbitCamMode then
				bin:Clean()
			end
		end)
	end
	function SpectateController:FitIndexToId()
		local _fn = math
		local _exp = self:GetSortedEntities()
		local _arg0 = function(e)
			return e.id == self.currentlySpectatingEntityId
		end
		-- ▼ ReadonlyArray.findIndex ▼
		local _result = -1
		for _i, _v in _exp do
			if _arg0(_v, _i - 1, _exp) == true then
				_result = _i - 1
				break
			end
		end
		-- ▲ ReadonlyArray.findIndex ▲
		self.spectateIndex = _fn.max(0, _result)
	end
	function SpectateController:GoToIncrement(mode, inc)
		local entities = self:GetSortedEntities()
		self.spectateIndex = (self.spectateIndex + inc) % #entities
		local entity = entities[self.spectateIndex + 1]
		self.currentlySpectatingEntityId = entity.id
		mode:SetTransform(entity.model.transform)
		ClientSignals.SpectatorTargetChanged:Fire({
			entity = entity,
		})
	end
	function SpectateController:GetSortedEntities()
		local team = Game.LocalPlayer:GetTeam()
		local _result
		if team then
			local _exp = self.bwController:GetAlivePlayersOnTeam(team)
			local _arg0 = function(player)
				return player.Character ~= nil
			end
			-- ▼ ReadonlyArray.filter ▼
			local _newValue = {}
			local _length = 0
			for _k, _v in _exp do
				if _arg0(_v, _k - 1, _exp) == true then
					_length += 1
					_newValue[_length] = _v
				end
			end
			-- ▲ ReadonlyArray.filter ▲
			local _arg0_1 = function(player)
				return player.Character
			end
			-- ▼ ReadonlyArray.map ▼
			local _newValue_1 = table.create(#_newValue)
			for _k, _v in _newValue do
				_newValue_1[_k] = _arg0_1(_v, _k - 1, _newValue)
			end
			-- ▲ ReadonlyArray.map ▲
			_result = _newValue_1
		else
			_result = self.entityController:GetEntities()
		end
		local entities = _result
		local _arg0 = function(a, b)
			return a.id < b.id
		end
		table.sort(entities, _arg0)
		return entities
	end
end
-- (Flamework) SpectateController metadata
Reflect.defineMetadata(SpectateController, "identifier", "Bundles/Client/Controllers/Match/SpectateController@SpectateController")
Reflect.defineMetadata(SpectateController, "flamework:parameters", { "Bundles/Client/Controllers/Global/Camera/CameraController@CameraController", "Bundles/Client/Controllers/Global/Entity/EntityController@EntityController", "Bundles/Client/Controllers/Match/BWController@BWController" })
Reflect.defineMetadata(SpectateController, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(SpectateController, "$:flamework@Controller", Controller, { {} })
return {
	SpectateController = SpectateController,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Client/Controllers/Match/SpectateController.ts
import { Controller, Dependency, OnStart } from "@easy-games/flamework-core";
import { ClientSignals } from "Client/ClientSignals";
import { CameraController } from "Imports/Core/Client/Controllers/Camera/CameraController";
import { OrbitCameraMode } from "Imports/Core/Client/Controllers/Camera/DefaultCameraModes/OrbitCameraMode";
import { LocalEntityController } from "Imports/Core/Client/Controllers/Character/LocalEntityController";
import { EntityController } from "Imports/Core/Client/Controllers/Entity/EntityController";
import { InventoryUIController } from "Imports/Core/Client/Controllers/Inventory/InventoryUIController";
import { PlayerController } from "Imports/Core/Client/Controllers/Player/PlayerController";
import { CoreClientSignals } from "Imports/Core/Client/CoreClientSignals";
import { Entity } from "Imports/Core/Shared/Entity/Entity";
import { Game } from "Imports/Core/Shared/Game";
import { Keyboard, Mouse } from "Imports/Core/Shared/UserInput";
import { Bin } from "Imports/Core/Shared/Util/Bin";
import { SetTimeout } from "Imports/Core/Shared/Util/Timer";
import { BWController } from "./BWController";

@Controller({})
export class SpectateController implements OnStart {
	public readonly spectateCamDistance = 5;

	private spectateIndex = 0;
	private currentlySpectatingEntityId = 0;

	constructor(
		private readonly cameraController: CameraController,
		private readonly entityController: EntityController,
		private readonly bwController: BWController,
	) {}

	OnStart() {
		// Start spectating once player is eliminated:
		ClientSignals.PlayerEliminated.Connect((event) => {
			if (event.player !== Game.LocalPlayer) return;
			SetTimeout(0.1, () => {
				this.StartSpectating();
			});
		});

		// const keyboard = new Keyboard();
		// keyboard.OnKeyDown(KeyCode.M, (event) => {
		// 	this.StartSpectating();
		// });
	}

	public ObserveSpectatorTarget(callback: (entity: Entity | undefined) => CleanupFunc): () => void {
		let cleanup: CleanupFunc;
		let found = false;
		if (this.currentlySpectatingEntityId) {
			const entity = Entity.FindById(this.currentlySpectatingEntityId);
			if (entity) {
				found = true;
				cleanup = callback(entity);
			}
		}
		if (!found) {
			cleanup = callback(undefined);
		}

		const bin = new Bin();

		bin.Add(
			ClientSignals.SpectatorTargetChanged.Connect((event) => {
				cleanup?.();
				cleanup = callback(event.entity);
			}),
		);

		return () => {
			print("cleaning.");
			bin.Clean();
		};
	}

	public StartSpectating() {
		Dependency<InventoryUIController>().SetEnabled(false);

		const entities = this.GetSortedEntities();
		if (entities.size() === 0) return;

		const bin = new Bin();

		const orbitCamMode = new OrbitCameraMode(this.spectateCamDistance, entities[0].model.transform);
		this.cameraController.SetMode(orbitCamMode);
		Dependency<LocalEntityController>().SetFirstPerson(false);

		this.GoToIncrement(orbitCamMode, 0);

		// Handle changing who is spectated:
		const mouse = bin.Add(new Mouse());
		const keyboard = bin.Add(new Keyboard());
		mouse.LeftDown.Connect(() => {
			this.GoToIncrement(orbitCamMode, keyboard.IsKeyDown(KeyCode.LeftShift) ? -1 : 1);
		});

		bin.Connect(CoreClientSignals.EntitySpawn, (event) => {
			this.FitIndexToId();
		});

		bin.Connect(CoreClientSignals.EntityDespawn, (entity) => {
			// If currently-spectated entity goes away, switch to the next entity:
			if (entity.id === this.currentlySpectatingEntityId) {
				this.FitIndexToId();
				this.GoToIncrement(orbitCamMode, 1);
			} else {
				this.FitIndexToId();
			}
		});

		// Clean up after camera mode is changed:
		bin.Connect(this.cameraController.cameraSystem.ModeChangedBegin, (newMode, oldMode) => {
			if (oldMode === orbitCamMode) {
				bin.Clean();
			}
		});
	}

	/** Adjust the index to reflect the currently-spectated ID. */
	private FitIndexToId() {
		this.spectateIndex = math.max(
			0,
			this.GetSortedEntities().findIndex((e) => e.id === this.currentlySpectatingEntityId),
		);
	}

	/** Increment the index and switch view to entity at that index. */
	private GoToIncrement(mode: OrbitCameraMode, inc: number) {
		const entities = this.GetSortedEntities();
		this.spectateIndex = (this.spectateIndex + inc) % entities.size();
		const entity = entities[this.spectateIndex];
		this.currentlySpectatingEntityId = entity.id;
		mode.SetTransform(entity.model.transform);
		ClientSignals.SpectatorTargetChanged.Fire({ entity });
	}

	/** Get a list of valid entities that can be spectated, sorted by ID. */
	private GetSortedEntities(): Entity[] {
		const team = Game.LocalPlayer.GetTeam();
		if (team) {
			const alivePlayers = this.bwController.GetAlivePlayersOnTeam(team).mapFiltered((p) => p.Character);
			if (alivePlayers.size() > 0) {
				return alivePlayers;
			}
		}
		// Allow spectating everyone when no more teammates are alive.
		return Dependency<PlayerController>()
			.GetPlayers()
			.filter((p) => !this.bwController.IsPlayerEliminated(p))
			.mapFiltered((p) => p.Character);

		// All entities
		// return this.entityController
		// 	.GetEntities()
		// 	.filter((e) => !e.IsLocalCharacter())
		// 	.sort((a, b) => a.id < b.id);
	}
}
